Adding typing

scirelli opened the pull request:
Adding `.venv` to the ignore list so it doesn't accidentally get added
Adding py.typed file to signify this library has typing information
Fixed typing errors found by mypy

Addresses issue #83

scirelli commented:
I'm not sure how to test the changes. Any advice?

scirelli commented:
Looks like black made some changes. Is there a way to rerun the pipeline?

tekktrik commented:
Nice!  I'll try to take a look at this soon, but maybe someone else will first!

tammymakesthings reviewed (commented):
These changes all look reasonable to me and I don’t see any problems. But I can’t find this board in my collection, so I personally would feel better if someone verified it on the hardware before merging.

tekktrik commented on adafruit_ads1x15/ads1015.py:
```diff
@@ -61,6 +61,5 @@ def _data_rate_default(self) -> int:
     def _conversion_value(self, raw_adc: int) -> int:
-        raw_adc = raw_adc.to_bytes(2, "big")
-        value = struct.unpack(">h", raw_adc)[0]
+        value = struct.unpack(">h", raw_adc.to_bytes(2, "big"))[0]
```
Was this combined for any particular reason?

tekktrik commented on adafruit_ads1x15/ads1015.py:
```diff
@@ -61,6 +61,5 @@ def _data_rate_default(self) -> int:
     def _conversion_value(self, raw_adc: int) -> int:
-        raw_adc = raw_adc.to_bytes(2, "big")
-        value = struct.unpack(">h", raw_adc)[0]
+        value = struct.unpack(">h", raw_adc.to_bytes(2, "big"))[0]
```
Was this (and the other similar change) combined for any particular reason?

scirelli replied:
`raw_adc` is type `int`, line 64 converts it to bytes, mypy complained. I didn't see a good reason to break it out so I combined the lines.

The other option would be to create a new variable.

tekktrik commented on adafruit_ads1x15/ads1x15.py:
```diff
@@ -15,11 +15,13 @@
 
 try:
     from typing import Optional
+
```
Are these new lines needed?

scirelli replied:
No. I think it was w.e pylint or pep linter i have installed. I can remove it.

scirelli replied:
Can someone explain this to me https://github.com/adafruit/Adafruit_CircuitPython_ADS1x15/pull/86/files#diff-ad1a84f70997c40a92f1b3d7688f0cbf2311c39d2e453f21b2c779f0f74d13c7R37

What is meant here? Gains index's are type float or int?

tekktrik replied:
The `gains` property returns a list of `float` since one of them is a fraction.

tekktrik commented on adafruit_ads1x15/analog_in.py:
```diff
@@ -12,7 +12,8 @@
 
 try:
-    from typing import Optional
+    from typing import Optional, cast
```
`cast()` cannot be used functionally, so it'll need to be removed.

tekktrik commented on .gitignore:
```diff
@@ -45,3 +45,4 @@ _build
 .idea
 .vscode
 *~
+.venv
```
I'm rolling out a patch that will put this under `.env`, so would you mind doing that?  Thanks!

tekktrik reviewed (changes requested):
Feedback below!  Also, would you mind adding type annotations for the properties getter and setters as well?  Thanks!

scirelli commented:
> These changes all look reasonable to me and I don’t see any problems. But I can’t find this board in my collection, so I personally would feel better if someone verified it on the hardware before merging.

I have a bbb and ads1115, but not an ads1015

tekktrik commented:
@scirelli, can you use the generics from the `typing` module like `Dict` and `List` instead of built-ins?  Using built-ins as generics wasn't available in Python 3.7.

tekktrik commented:
@scirelli let me know when I should re-review

scirelli commented:
I have not had a chance to test on actual hardware. Other than that I addressed all PR comments.

Which tests should I run? I was looking at the simple one in examples.

tekktrik commented on .gitignore:
```diff
@@ -45,3 +45,4 @@ _build
 .idea
 .vscode
 *~
+.env
```
This is a duplicate, I think you meant to add `.venv`.  Can you add it underneath `.env` too?  It'll match a patch I'm hoping to roll out.

scirelli replied:
Oh I misunderstood what you were asking. I thought you wanted me to rename my local version and add `.env`

tekktrik replied:
I'll re-review and add this to that.

tekktrik commented:
I also realize that I haven't reviewed the property getters and setters, I'll do that in a few hours.  Sorry for the delay!

tekktrik commented on adafruit_ads1x15/ads1x15.py:
```diff
@@ -69,66 +71,73 @@ def __init__(
         self._last_pin_read = None
         self.buf = bytearray(3)
-        self._data_rate = self._gain = self._mode = None
+        self._gain: float
```
Can you remove these annotations?  Since the CircuitPython parser sheds them, I'm worried this might look like a regular statement, so it will look like it hasn't been defined yet.  This goes for `self._data_rate` and `self._mode`.

tekktrik replied:
You could test this general premise of type annotating something like this, though, in a CircuitPython driver.  Let me know if you can and if it works!

scirelli replied:
Should I just remove them or initialize them to something and let the property setters set them again based on their logic?
```
        self._gain = gain
        self.gain = gain
        self._data_rate = self._data_rate_default() if data_rate is None else data_rate
        self.data_rate = self._data_rate_default() if data_rate is None else data_rate
        self._mode = mode
        self.mode = mode
```

tekktrik replied:
See next review comment!

scirelli commented on adafruit_ads1x15/ads1115.py:
```diff
@@ -42,26 +43,25 @@ class ADS1115(ADS1x15):
         """Rate configuration masks."""
         return _ADS1115_CONFIG_DR
 
     def _data_rate_default(self) -> int:
```
It's returning the entire Dict not a value in it. If I change it the types won't match.

tekktrik replied:
This is referring to the new Line 62.

scirelli replied:
I'm blind

tekktrik commented on adafruit_ads1x15/ads1x15.py:
```diff
@@ -69,66 +71,73 @@ def __init__(
         self._last_pin_read = None
         self.buf = bytearray(3)
-        self._data_rate = self._gain = self._mode = None
+        self._gain = gain
```
You should be able to just use the setters, which set the underlying hidden variables themselves:

```python
self.gain = gain
```

Basically, from the original, the only change needed (unless something like pylint gets angry) is to remove setting everything to `None` prior.  If it does get upset, honestly setting them to `None` isn't bad - they'll all get overwritten in the setters, and given the implementation that nothing will initialize if valid values aren't selected, there isn't a need to type any of the getters as `Optional`.

tekktrik replied:
Feel free to request a review after address this so I don't block you :)

tekktrik commented on adafruit_ads1x15/ads1x15.py:
```diff
@@ -69,66 +71,70 @@ def __init__(
 
     @property
-    def data_rate(self):
+    def bits(self) -> Union[Literal[12], Literal[16]]:
```
Since, you're saying the subclass should implement this, let's type it as an `int` return to say that we're promising that bits will return an integer, and we can be more specific as we have in said subclasses.

tekktrik reviewed (approved):
Great work!

scirelli commented:
Thanks! I learned a few things, still new to Python typing.
